### PR TITLE
Feature/bst

### PR DESCRIPTION
**Pull request by sjschmidt44**

### CODE REVIEW

File: bst.py [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+            return self.right.contains(val)
+
+    def size(self):
+        """Return the total number of nodes in the tree.
+
+        returns: integer of total node; 0 if empty
+        """
+        left_size = self.left.size() if self.left is not None else 0
+        right_size = self.right.size() if self.right is not None else 0
+        return left_size + right_size + 1
```
Comment (cewing) [inline bst.py]: very good.

File: bst.py [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+        returns: integer
+        """
+        left_depth = self.left.depth() if self.left is not None else 0
+        right_depth = self.right.depth() if self.right is not None else 0
+        if left_depth > right_depth:
+            return 1
+        elif left_depth < right_depth:
+            return -1
+        else:
+            return 0
```
Comment (cewing) [inline bst.py]: No.  we actually want to have the degree of imbalance.  Not a simple 1 or -1, but a reflection of how out of balance things actually are.

File: bst.py [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+    """
+    worst = Node(1)
+    for val in range(2, 32):
+        worst.insert(val)
+
+    best = Node(16)
+    best_values = [
+        8, 24, 4, 12, 20, 28, 2, 6, 10, 14, 18, 22, 26, 30, 1, 3, 5,
+        7, 9, 11, 13, 15, 17, 19, 21, 23, 25, 27, 29, 31
+    ]
```
Comment (cewing) [inline bst.py]: There must be a better way to do this, don't you think?

File: bst.py [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+    for val in best_values:
+        best.insert(val)
+
+    worst_case = Timer(
+        'worst.contains(31)', 'from __main__ import worst').timeit(1000)
+    best_case = Timer(
+        'best.contains(31)', 'from __main__ import best').timeit(1000)
+
+    print "The worst case took {}.".format(worst_case)
+    print "The best case took {}.".format(best_case)
```
Comment (cewing) [inline bst.py]: given the small size of your trees, I don't imagine this difference is very large.

File: test_bst.py [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+    root.insert(14)
+
+    return root
+
+
+def test_insert(rand_setup):
+    pre_size = rand_setup.size()
+    new = 200
+    rand_setup.insert(new)
+    assert rand_setup.insert(new) is None
```
Comment (cewing) [inline test_bst.py]: what does this assertion prove?

File: test_bst.py [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+    rand_setup.insert(new)
+    post_size = rand_setup.size()
+    assert post_size > pre_size
+    assert post_size == pre_size + 1
+
+
+def test_contains(rand_setup):
+    rand = randint(1, 100)
+    rand_setup.insert(rand)
+    assert rand_setup.contains(rand) is True
```
Comment (cewing) [inline test_bst.py]: This is a pretty minimal test, don't you think?

File: test_bst.py [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+
+
+def test_size(rand_setup):
+    pre_size = rand_setup.size()
+    new = 200
+    rand_setup.insert(new)
+    rand_setup.insert(new)
```
Comment (cewing) [inline test_bst.py]: if `insert` is broken, what's the impact on this test?

File: test_bst.py [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+    root.insert(12)
+    root.insert(24)
+    root.insert(40)
+    root.insert(14)
+
+    return root
+
+
+def test_insert(rand_setup):
+    pre_size = rand_setup.size()
```
Comment (cewing) [inline test_bst.py]: What if `size` isn't working correctly?

File: test_bst.py [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+
+def test_balance(rand_setup):
+    left = rand_setup.left.depth() if rand_setup.left is not None else 0
+    right = rand_setup.right.depth() if rand_setup.right is not None else 0
+    if left > right:
+        assert rand_setup.balance() == 1
+    elif right > left:
+        assert rand_setup.balance() == -1
+    else:
+        assert rand_setup.balance() == 0
```
Comment (cewing) [inline test_bst.py]: this is pretty much a tautology.  It just proves you can type correctly.

File: bst.py (hunk 2 above)
Comment (jonathanstallings) [inline bst.py]: Oh my. I misinterpreted the spec there. Thank you for the clarification. I will fix this.

**Comment by jonathanstallings** [comment]
Thank you for your thorough feeback, Cris.

**Pull request merged**